### PR TITLE
WIP: Add 'options' argument in mkProxy(), add specific interfaces and allo…

### PR DESCRIPTION
**Pull request by FredericAMAN**
…w to add new interfaces when specific proxy (#33)

### CODE REVIEW

File: lib/DBusProxy.js [hunk 1]
```diff
@@ -140,7 +143,9 @@ function DBusProxy (serviceName = mandatory(), targetedPath, bus = mandatory(),
 
 		this.emit('connected')
 	})
-
+	.then( _ => {
+		debug(`Introspection done, this:\n${inspect(this, {depth: 7})}`)
```
Comment (antoineDeRengerve) [inline lib/DBusProxy.js:147]: Should debug message remain ? With depth is 7 ? Is it always 7 ?

File: lib/DBusProxy.js [hunk 2]
```diff
@@ -294,13 +298,24 @@ DBusProxy.prototype.makeIntrospectionPass = function (introspectionDepth, target
...
+
+				if (targetedIface != null) {
+					if ((Array.isArray(targetedIface) && targetedIface.indexOf(iface.attr('name').value()) === -1) || (typeof targetedIface === 'string' && targetedIface !== iface.attr('name').value())) {
+						debug(`In '${path}', '${iface.attr('name').value()}' doesn't match '[${targetedIface}]'`)
+						continue
+					} else {
+						debug(`In '${path}', '${iface.attr('name').value()}' matches '[${targetedIface}]'`)
+					}
+				} else {
+					debug(`In '${path}', no targeted interface`)
```
Comment (antoineDeRengerve) [inline lib/DBusProxy.js]: Actually targetting all interfaces because they will be all loaded

File: lib/DBusProxy.js [hunk 3]
```diff
@@ -685,6 +698,18 @@ DBusProxy.prototype.makeIntrospectionPass = function (introspectionDepth, target
 	}
 }
 
+DBusProxy.prototype.getDepth = function (objectPath) {
+	let objectPathSplit = objectPath.split('/')
+	return objectPathSplit.length
+}
+
+DBusProxy.prototype.pathsAreIncluded = function (path1, path2) {
```
Comment (antoineDeRengerve) [inline lib/DBusProxy.js]: Comment function

File: lib/DBusProxy.js [hunk 4]
```diff
@@ -642,28 +660,23 @@ DBusProxy.prototype.makeIntrospectionPass = function (introspectionDepth, target
...
-						if (!(targetedPath.indexOf(newPath) !== -1 || newPath.indexOf(targetedPath) !== -1)) {
-							debug(`${newPath} doesn't match ${targetedPath}`)
-							continue
-						}
-					} else {
-						debug("No targeted object path")
+					if (!this.pathsAreIncluded(targetedPath, newPath)) {
+						debug(`'${newPath}' doesn't match '${targetedPath}'`)
+						continue
+					} else if (newDepth > 0) {
```
Comment (antoineDeRengerve) [inline lib/DBusProxy.js]: Add a comment : why newDepth must be >0

File: lib/DBusProxy.js [hunk 5]
```diff
@@ -571,34 +586,37 @@ DBusProxy.prototype.makeIntrospectionPass = function (introspectionDepth, target
...
-										.then( _ => {
-											debug(`done, this:\n${inspect(this, {depth: 7})}`)
-										})
 									}
 
 									/*
 										Handle 'InterfacesRemoved' signal.
 									*/
-									if (signalName === 'InterfacesRemoved' && this.isInterfaceUpdatingAllowed) {
+									if (signalName === 'InterfacesRemoved') {
```
Comment (antoineDeRengerve) [inline lib/DBusProxy.js:619]: What if an interface that was not managed is removed ? Does it make the library crash ?
Have you tested ?

**Review by antoineDeRengerve — CHANGES_REQUESTED** [review]
Some changes expected for maintainability : comments, etc.

One real question about managing interfaceRemoved when not all interfaces/objects are loaded.

**Comment by FredericAMAN** [comment]
@antoineDeRengerve I did some modifications according to your comments

File: lib/DBusProxy.js (hunk 1 above)
Comment (antoineDeRengerve) [reply lib/DBusProxy.js:147]: So ?

File: lib/DBusProxy.js (hunk 5 above)
Comment (antoineDeRengerve) [reply lib/DBusProxy.js:619]: The change seems ok with respect to this question. I assume, you have tested.

**Comment by antoineDeRengerve** [comment]
Please provide replies to remaining questions.

File: lib/DBusProxy.js (hunk 1 above)
Comment (FredericAMAN) [reply lib/DBusProxy.js:147]: Yes it should, to help for debugging by displaying the final proxy object. By default, introspection depth is 2: 7 is to show deeper

**Comment by FredericAMAN** [comment]
Ok, it's done

**Review by antoineDeRengerve — APPROVED** [review]

**Pull request merged**